core/src/net/network_thread.rs

```diff
@@ -769,14 +771,46 @@ impl NetworkThread {
         }
         let rc = Rc::new(RefCell::new(channel));
         self.address_map.insert(address, rc.clone());
-        self.token_map.insert(self.token, rc);
+        self.token_map.put(self.token, rc);
+        self.check_connection_limit();
         self.next_token();
     }
 
-    /// Initiates a graceful closing sequence
+    /// Checks the current channel-count and initiates a graceful shutdown of the channel.
+    fn check_connection_limit(&mut self) -> () {
+        let channel_count = self.token_map.len() as i32;
+        let mut exceeded = channel_count - self.network_config.get_connection_limit() as i32;
+        while exceeded > 0 {
+            let mut lru_iter = self.token_map.iter().rev();
+            let addr;
+            if let Some((_, channel)) = lru_iter.next() {
+                warn!(
+                    self.log,
+                    "Connection Limit exceeded! {} concurrent channels, limit = {}. \
+                Closing channel {:?}",
+                    channel_count,
+                    self.network_config.get_connection_limit(),
+                    channel.borrow(),
+                );
+                self.notify_network_event(NetworkEvent::ConnectionLimitExceeded());
+                addr = channel.borrow().address();
+                self.close_connection(addr);
+                exceeded -= 1;
+            }
+        }
+    }
+
+    /// Initiates a graceful closing sequence if the channel is connected or
     fn close_connection(&mut self, addr: SocketAddr) -> () {
         if let Some(channel) = self.get_channel_by_address(&addr) {
-            let _ = channel.borrow_mut().initiate_graceful_shutdown();
+            let mut channel_mut = channel.borrow_mut();
+            if channel_mut.connected() {
+                channel_mut.initiate_graceful_shutdown();
+                // Update LRU position
+                let _ = self.get_channel_by_token(&channel_mut.token);
```

Comment:
Oh, I see now...we are closing stuff asynchronously, but we are already allowing the new connections.

I'm not sure this is a good idea, tbh. This still opens us up to potential crashes due to DoS. I'd prefer us maybe having a soft limit using this procedure (or something similar) and a hard limit where we just immediately reject incoming connections until there is space again.